src/Features/CSharp/Portable/RemoveUnreachableCode/CSharpRemoveUnreachableCodeDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,161 @@
+            //
+            // To avoid this, we simply ask the semantic model for all the diagnostics for this
+            // block and we look for any reported "unreachable code detected" diagnostics.
+            //
+            // This is actually quite fast to do because the compiler does not actually need to
+            // recompile things to determine the diagnostics.  It will have already stashed the
+            // binding diagnostics directly on the SourceMethodSymbol containing this block, and
+            // so it can retrieve the diagnostics at practically no cost.
+            var root = semanticModel.SyntaxTree.GetRoot(cancellationToken);
+            var diagnostics = semanticModel.GetDiagnostics(block.Span, cancellationToken);
+            foreach (var diagnostic in diagnostics)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (diagnostic.Id == CS0162)
+                {
+                    ProcessUnreachableDiagnostic(context, root, diagnostic.Location.SourceSpan);
+                }
+            }
+        }
+
+        private bool IsOutermostBlock(SyntaxNode block)
+        {
+            // Avoid linq/allocations as we will be calling this on every block in the solution.
+            for (var current = block.Parent; current != null; current = current.Parent)
+            {
+                if (current.Kind() == SyntaxKind.Block)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private void ProcessUnreachableDiagnostic(
+            SyntaxNodeAnalysisContext context, SyntaxNode root, TextSpan sourceSpan)
+        {
+            var node = root.FindNode(sourceSpan);
+
+            // Note: this approach works as the language only supports the concept of 
+            // unreachable statements.  If we ever get unreachable subexpressions, then
+            // we'll need to revise this code accordingly.
+            var firstUnreachableStatement = node.FirstAncestorOrSelf<StatementSyntax>();
+            if (firstUnreachableStatement == null ||
+                firstUnreachableStatement.SpanStart != sourceSpan.Start)
+            {
+                return;
+            }
+
+            // At a high level, we can think about us wanting to fade out a "section" of unreachable
+            // statements.  However, the compiler only reports the first statement in that "section".
+            // We want to figure out what other statements are in that section and fade them all out
+            // along with the first statement.  This is made somewhat tricky due to the fact that
+            // subsequent sibling statements possibly being reachable due to explicit gotos+labels.
+            //
+            // On top of this, an unreachable section might not be contiguous.  This is possible 
+            // when there is unreachable code that contains a local function declaration in-situ.
```

Comment:
> then it is not reachable. 

I feel like we're putting on pedantry hats here :)   Being totally pedantic, i would say that it's totally reachable, it's just not referenced.  Regardless, i'm not sure anything needs to be done here.  I've ensured the code works properly in the case of actually referenced local functions, and i've opened a bug to track ensuring that users can easily remove unreferenced local functions to compliment the work we've done to let them remove unreferenced local variables.  I'm satisfied with that.